src/constructRoutes.js

```diff
@@ -6,10 +6,37 @@ import {
   validateObject,
 } from "./validation-helpers.js";
 
+/**
+ * @typedef {{
+ * mode?: string;
+ * base?: string;
+ * containerEl?: string | HTMLElement;
+ * disableWarnings?: boolean;
+ * routes: Array<Route>;
+ * }} RoutesConfig
+ *
+ * @typedef {UrlRoute | Application} Route
+ *
+ * @typedef {{
+ * type: 'route';
+ * path: string;
+ * routes?: Array<Route>;
+ * }} UrlRoute
+ *
+ * @typedef {{
+ * type: 'application';
+ * name: string;
+ * props: object;
+ * }} Application
+ *
+ * @param {RoutesConfig} routesConfig
+ * @returns {any}
+ */
```

Comment:
Potentially dumb question:

Do we want to have types in the project or use something like absolutely typed?

I saw the discussion with Anthony in slack and I think this is actually JSDOC and types in one so I can see an argument for having them here.

Reply:
I'm not sure what absolutely typed is - are you talking about DefinitelyTyped?

I know that you really don't like typescript, but a lot of people do. I have done all of the following: manually  maintained type defs in the repo (single-spa), manually maintained type defs in DefinitelyTyped repo, and automatically maintained type defs from a typescript project (OpenMRS). I have never done this in-between approach of sorta-automatically maintained via jsdoc comments, and would like to try it out to see if it's a good in-between step or not.

Reply:
Yeah definitelyTyped sorry.

I'm not a huge typescript fan, but that's not where my question is coming from. I am happy to try this approach, I was hoping to understand the pros/cons better of the various options.

I've never tried maintaining types in any of the ways you mentioned so the question I should have asked is:

This seems like an interesting approach, what are the advantages to doing it with JSDOC vs DefinitelyTyped? I assume there would be several but there is a lot I don't understand.

Reply:
When using DefinitelyTyped:

1. User has to do `npm install --save @types/single-spa-layout` in order to have the types
1. Types are not colocated with the code that defines them, which makes it harder to keep in sync.
1. Types are published separately from the project itself, which means they have a different version on npm. Users must find which version of DefinitelyTyped matches the version of single-spa-layout
1. IDEs do not provide up-to-date autocomplete for people working on single-spa-layout itself, since the autocomplete is either non-existent (by default) or stale (when single-spa-layout installs `@types/single-spa-layout`
1. You do have the full power of typescript

When using jsdoc typedefs:

1. Types come with the package - no need to install `@types/single-spa-layout` npm packakge
1. Types are colocated with the code
1. Types share a version with the project, so there's no syncing/versioning issue with `@types/single-spa-layout`
1. IDEs provide autocomplete for contributors to single-spa-layout
1. You are limited by typescript's support of jsdocs, which is not fully expressive of everything I think. See https://stackoverflow.com/questions/37355244/ignore-cannot-find-module-error-on-typescript